Add kafka-connector blog

martindekov opened the pull request:
Adding blog post which guides
how to use the kafka-connector

Closes #76 

Signed-off-by: Martin Dekov <mdekov@vmware.com>

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,183 @@
+
+## Kafka Connector
+
+Kafka Connector plugin is the solution to your problem when it comes to Kafka. You just need to apply annotation with the topic you would like to subscribe your function to and then run the connector with your Kafka configuration.
```
What is my problem with Kafka?

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,183 @@
+
+If you deployed your OpenFaaS instance with Helm and default values there is no need to make changes to this file. The `kafka_broker` variable is our broker host from which we will pull topics.
+
+In our case the broker host is located in service called `kafka` and it is in the same namespace as the gateway plus it uses the default kafka port `9092`.
```
'In our case the broker is located within a service called `kafka`, which resides in the same namespace as the OpenFaaS gateway and uses the default kafka port of `9092`.'

martindekov replied:
Missed those comments, now they are referenced 👍

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,183 @@
+
+In our case the broker host is located in service called `kafka` and it is in the same namespace as the gateway plus it uses the default kafka port `9092`.
+
+The environmental variable `topics` is the place in which we list the topics in which we are interested to receive from the broker.
```
We are talking in the plural but only show one topic in the code block.  Should we elaborate on how multiple topics are listed?

martindekov replied:
Added examples on how to use multiple topics ✔️

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
````diff
@@ -0,0 +1,183 @@
+$ kubectl apply -f yaml/kubernetes
+```
+
+We will deploy simple producer provided by Alex:
````
Why are we doing this?  Add context

martindekov replied:
Context added 👍

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,183 @@
+
+## Subscribe function
+
+In order to consume topics from the connector we should just apply annotation with key `topic` and value whichever you used in the `connector-dep.yml` in our case it was `faas-request`.
```
In order to consume topics via the connector we simply need to apply an annotation with a key of `topic` and the chosen value used in the `connector-dep.yml` - in our case it was `faas-request`.

What if there are multiple topics stated in `connector-dep.yml`?  Can a function subscribe to multiple topics?

martindekov replied:
Added how to subscribe to multiple topics even though explicitly it is not supported 👍

rgee0 replied:
Just to capture what we discussed on Slack, and for the benefit of others viewing this, we've established that observing multiple topics via the connector is a supported feature.

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
````diff
@@ -0,0 +1,183 @@
+$ kubectl logs $(kubectl get pods -o name -n openfaas | grep -m1 kafka-connector | cut -d'/' -f 2) -n openfaas --tail 100 -f
+```
+
+In the first terminal, inside the producer pod shell session invoke the producer binary with proper command line arguments:
````
Proper?

martindekov replied:
Swapped with `those` ✔️

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
````diff
@@ -0,0 +1,183 @@
+/usr/bin producer -messages=25 -broker=kafka:9092 -pause 1s -topic=faas-request
+```
+
+You should see the output of the function in the kafka-connector logs.
````
Lets code block some output?

martindekov replied:
Added sample output 👍

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,183 @@
+
+You should see the output of the function in the kafka-connector logs.
+
+## Wrapping up
```
I think we need a bit more here to stimulate people's thinking.  A real world use case perhaps?  Online order triggering billing , invoicing, and warehousing functions?

martindekov replied:
I will spend some time how to formulate those, good proposal 👍

martindekov replied:
Added paragraph before that one showing the real life example 💯

martindekov commented:
Thanks for the review @rgee0 I will reference your proposals tommorow 🍻

martindekov commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,183 @@
+---
+title: "Subscribe your OpenFaaS functions to Kafka"
```
Done 👍

martindekov commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,183 @@
+dark_background: true
+---
+
+In this post I will show you how to subscribe your OpenFaaS functions to topics in Kubernetes.
```
Sounds better ^ ✔️

martindekov commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,183 @@
+
+In this post I will show you how to subscribe your OpenFaaS functions to topics in Kubernetes.
+
+OpenFaaS is production ready Serverless framework, but often times it finds itself as a part of bigger eco-system of software and the different bits of that eco-system may communicate with each-other in different ways. The OpenFaaS gateway exposes REST endpoint, but what if our services don't use HTTP?
```
Attention to detail, nice 😄 Done 👍

martindekov commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,183 @@
+In this post I will show you how to subscribe your OpenFaaS functions to topics in Kubernetes.
+
+OpenFaaS is production ready Serverless framework, but often times it finds itself as a part of bigger eco-system of software and the different bits of that eco-system may communicate with each-other in different ways. The OpenFaaS gateway exposes REST endpoint, but what if our services don't use HTTP?
+
```
Good way to trigger my thinking 💯 Done 👍

alexellis commented:
I would like us to figure out / sort-out the helm chart before we go live. Do we think that's possible?

alexellis commented on _posts/2019-03-14-2019-kafka-connector.md:
````diff
@@ -0,0 +1,217 @@
+$ cd kafka-connector/yaml/kubernetes
+```
+
+The file `connector-dep.yaml` contains all the configuration you need to set for your connector.
````
With helm this gets a lot simpler.

The topics section would be passed via a helm chart option.

The Kafka itself would be deployed from a Kafka chart and not from our chart.

martindekov replied:
I think we discussed somewhere that the Kafka stack provided by wurstmeister does not have Charts repository. I will add installation through Charts for the Connector though

martindekov replied:
Also can't seem to add list to key using Helm's `--set` flag. The output `... --set-string topics=topic1,topic2` gives me 
```
Error: failed parsing --set-string data: key "whatever2" has no value
```

martindekov replied:
This part is done 👍

alexellis commented on _posts/2019-03-14-2019-kafka-connector.md:
````diff
@@ -0,0 +1,217 @@
+          secretName: basic-auth
+```
+
+If you deployed your OpenFaaS instance with Helm and default values there is no need to make changes to this file. The `kafka_broker` variable is our broker host from which we will pull topics.
````
We need to make sure people know how to connect to their existing broker including use of the namespace.

martindekov replied:
Added comment that if their kafka service is in another namespace they can append `.<namespace>` to make the broker discoverable by the connector 💯

alexellis commented on _posts/2019-03-14-2019-kafka-connector.md:
````diff
@@ -0,0 +1,217 @@
+Before we proceed, make sure you have all the components running:
+
+```bash
+$ kubectl get pods -n openfaas
````
Please make this just look for the specific parts related to the post.

martindekov replied:
Done 👍 
```
kubectl get pods -n openfaas | grep -E 'kafka|zookeeper|producer'
kafka-broker-544fbccd48-p6pgr     1/1     Running            3          71m
kafka-connector-5d9f447f5-5drv7   1/1     Running            0          69m
zookeeper-699f568f6f-6b4n2        1/1     Running            0          71m
```

alexellis commented on _posts/2019-03-14-2019-kafka-connector.md:
````diff
@@ -0,0 +1,217 @@
+We will use function from the store called `figlet`:
+
+```bash
+$ faas-cli store deploy figlet --annotation "topic=faas-request"
````
I like this example, but I think we need an example with a YAML file too and a custom function.

martindekov replied:
Done 👍

alexellis commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,217 @@
+
+Now you might ask:
+
+"Ok, cool but how am I supposed see the output of my function?"
```
Ok, cool but  -> Great, but

martindekov replied:
Nice ✔️

alexellis commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,217 @@
+
+## Wrapping up
+
+Kafka Connector implements the [Connector SDK](https://github.com/openfaas-incubator/connector-sdk) with which you can create your own invoking mechanism. 
```
Nice reference. 

Link to the docs page on triggers too?

martindekov replied:
Done 🥇

alexellis commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,217 @@
+
+You can check the existing connectors like the [Cron Connector](https://github.com/zeerorg/cron-connector) or [vCenter Connector](https://github.com/openfaas-incubator/vcenter-connector).
+
+Now you are ready to attach the Kafka Connector plugin to your own Kafka deployment.
```
Final call to action.. checkout my latest posts for an example.

martindekov replied:
Prompted to extend the Hello world example and try multiple topics ✔️

alexellis commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,217 @@
+
+OpenFaaS is production ready Serverless framework, and often finds itself as a part of bigger eco-system of software where the disparate parts may communicate with each-other in different ways. The OpenFaaS gateway exposes REST endpoints, but what if our services don't use HTTP?
+
+Kafka is stream-processing platform and at a high level the important concepts are Producers, Consumers and Brokers which communicate with each-other via massages attached to Topics. The way it works is that the Producer is sending messages on specific Topics to the Broker, then the Broker notifies all the Consumers which are interested in that Topic. The approach is fitting for distributed syste ...
```
Link to Kafka docs?

martindekov replied:
Done 👍

embano1 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,217 @@
+
+OpenFaaS is production ready Serverless framework, and often finds itself as a part of bigger eco-system of software where the disparate parts may communicate with each-other in different ways. The OpenFaaS gateway exposes REST endpoints, but what if our services don't use HTTP?
+
+Kafka is stream-processing platform and at a high level the important concepts are Producers, Consumers and Brokers which communicate with each-other via massages attached to Topics. The way it works is that the Producer is sending messages on specific Topics to the Broker, then the Broker notifies all the Consumers which are interested in that Topic. The approach is fitting for distributed syste ...
```
> then the Broker notifies all the Consumers which are interested in that Topic
Nit pick: technically, the broker does not notify the consumers like in a pub/sub system. Consumers (long) poll the topics.

martindekov replied:
Thanks Michael 👍 Added that the connector actually polls the topics

rgee0 commented:
@martindekov a couple of images we should look to add while replacing the original photo

![image](https://user-images.githubusercontent.com/16418793/55188638-c6de2380-5193-11e9-94f5-fcd1ecbd1689.png)

![image](https://user-images.githubusercontent.com/16418793/55188662-d52c3f80-5193-11e9-98c1-8ce0f347a5e3.png)

martindekov commented:
Hey @rgee0 I decided to add the second picture in place of the old one as it is more detailed, and they both say basically the same thing with different level of detail

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,199 @@
+
+OpenFaaS is production ready Serverless framework, and often finds itself as a part of bigger eco-system of software where the disparate parts may communicate with each-other in different ways. The OpenFaaS gateway exposes REST endpoints, but what if our services don't use HTTP?
+
+Kafka is stream-processing platform and at a high level the important concepts are Producers, Consumers and Brokers which communicate with each-other via massages attached to Topics. The way it works is that the Producer is sending messages on specific Topics to the Broker, then the Consumers poll all the topics of interest from the Brokers. The approach is fitting for distributed systems, becaus ...
```
via massages => via messages

martindekov replied:
😄 Yeah nice

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,199 @@
+
+## Kafka Connector
+
+Kafka Connector plugin is the solution to your problem when it comes to integrating OpenFaaS as a component of a bigger distributed system which is using Kafka to make communication between its parts. You just need to apply annotation with the topic you would like to subscribe your function to and then run the connector with your Kafka configuration.
```
Start this section with a `The`

martindekov replied:
Changed

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,194 @@
+
+Here is a diagram which shows how the component works:
+
+![Form Options](/images/kafka-connector/kafka-connector-detailed-overview.png)
```
What is `Form Options`?

martindekov replied:
Adds the picture, at least that's how other blogs do it took it from there

martindekov replied:
changed to Kafka Diagram

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,194 @@
+
+## Pre-requisites
+
+First and most important thing is [running OpenFaaS instance](https://docs.openfaas.com/deployment/kubernetes/) with default configuration on Kubernetes, for our case [minikube deployment](https://kubernetes.io/docs/setup/minikube/) is more than enough.
```
The first

martindekov replied:
Done ✔️

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,194 @@
+
+## Deploy the components with configuration
+
+In order to deploy the Zookeeper and Broker apply the kubernetes files while inside the kubernetes folder from above:
```
apply the `yaml` files while in the `kubernetes` folder

martindekov replied:
Done 🥇

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,194 @@
+
+It is important to note that those files deploy the Zookeeper and Broker in the same namespace as the OpenFaaS Gateway.
+
+We will use charts to deploy the Kafka Connector. If you don't have the OpenFaaS charts repository, then add it with this command:
```
Helm charts

Do we need to link to the official 'how to install Helm' in case the user inst yet a helm user?

martindekov replied:
Done ✔️

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
````diff
@@ -0,0 +1,194 @@
+$ helm repo add openfaas https://openfaas.github.io/faas-netes/
+```
+
+Install Kafka Connector with default values:
````
install the

martindekov replied:
Done ✔️

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,194 @@
+    --namespace openfaas \
+    --set topics="faas-request\," \
+    --set broker_host="kafka" \
+    --set print_response_body="true"
```
`--set print_response_body="true"` wont be applied - it doesn't exist in `values.yml`

martindekov replied:
Added [PR](https://github.com/openfaas/faas-netes/pull/408)

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
````diff
@@ -0,0 +1,194 @@
+Before we proceed, make sure you have all the components running:
+
+```bash
+$ kubectl get pods -n openfaas | grep -E 'kafka|zookeeper|producer'
````
why do we grep for producer?

martindekov replied:
Alex requested in the upper comments to be more specific with the parts

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,194 @@
+
+In order to consume topics via the connector we simply need to apply an annotation with a key of `topic` and the chosen value used in the `connector-dep.yml` - in our case it was `faas-request`.
+
+Here it is important to note that if you would like your function to subscribe to multiple topics, you just need to list them as we did in the connector configuration.
```
The connector can bind to multiple topics, but functions cannot yet.

martindekov replied:
Should I remove this? Or we will merge it before the blog is out ? I was hoping to get it in the project before publishing the blog

martindekov replied:
Removed the reference

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,194 @@
+
+"Great, but how am I supposed see the output of my function?"
+
+Remember in the configuration we set `print_response_body` to `true` so we can see the response body in the logs of the Kafka Connector.
```
We cant, not yet

martindekov replied:
I hope to fix this before the blog is out but I can remove it ?

martindekov replied:
Removed the reference

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,194 @@
+
+There are multiple real world examples where Kafka in combination with the Serverless approach provided by OpenFaaS becomes really powerful tool as Kafka can make even the smallest of systems extensible and OpenFaaS provides us with the quick way to add those extensions.
+
+Lets say you need to process customer information. When customer applies for you application you need to process it and store the information in a database and there are tons of customers. Just create topic for example `user-creation` which will be populated with the information and then create OpenFaaS function which handles the processing and updating of that information to the database.
```
Not sure about the example, I've altered a little, but I'm still not 100% with it:

When a customer creates an account you need to process their information and store it in a database.  However, there are a number of other departments also interested in new customers - theres marketing who want to send a welcome pack, there's payments who want to check their credit score. Just create topic, for example `user-creation`, which will be populated with the information and then create OpenFaaS function which handles the processing and updating of that information to the database.  Similarly marketing and payments can both implement functions to process the data in a way that makes sense for them.

martindekov replied:
Thanks for this Richard I had problems coming up with good example

martindekov replied:
Changed the example with yours 💯

martindekov replied:
It is better

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,194 @@
+
+Lets say you need to process customer information. When customer applies for you application you need to process it and store the information in a database and there are tons of customers. Just create topic for example `user-creation` which will be populated with the information and then create OpenFaaS function which handles the processing and updating of that information to the database.
+
+If we extend on the example above, again you would like to add some quick feature for the user.Attach another function to the same topic to extend the already existing functionality or just add another topic and write the function which handles that feature, it is as simple as that. 
```
a space at `user.Attach`

martindekov replied:
Done ✔️

rgee0 commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,194 @@
+
+You can check the existing connectors like the [Cron Connector](https://github.com/zeerorg/cron-connector) or [vCenter Connector](https://github.com/openfaas-incubator/vcenter-connector) or all the existing function [Triggers](https://docs.openfaas.com/reference/triggers/).
+
+Now you already know the basics. It is over to you to deploy the connector with development tools like we did in the blog, or attach it to your existing Kafka deployment and start invoking functions. You can perhaps try attaching your function to multiple topics and extend the Hello World function with your own language of choice. Serverless with Kafka is all in your hands with the Kafka Connecto ...
```
You cant attach a function to multiple topics yet

martindekov replied:
I was hoping to merge this before the blog post is out?

martindekov replied:
Removed the reference

martindekov commented on _posts/2019-03-14-2019-kafka-connector.md:
```diff
@@ -0,0 +1,194 @@
+  | |/ _ \/ __| __| | __| '_ \ / _ \ | |_| | | | '_ \ / __| __| |/ _ \| '_ \   
+  | |  __/\__ \ |_  | |_| | | |  __/ |  _| |_| | | | | (__| |_| | (_) | | | |_ 
+  |_|\___||___/\__|  \__|_| |_|\___| |_|  \__,_|_| |_|\___|\__|_|\___/|_| |_(_)
+                                                                               
```
This should be changed with the real output once I get it working

martindekov replied:
Changed

martindekov commented:
Squashed the commits

rgee0 commented:
https://twitter.com/alexellisuk/status/1114080122733322240

martindekov commented:
Hey @rgee0 I used that image in the blog

alexellis commented:
@embano1 please can you give the updated version a read-through to see if your feedback has been addressed?

Thanks

embano1 commented:
LGTM

martindekov commented:
@alexellis Compressed the image to 303 kb and with 1600xAutovalue. Also changed the date to `04-17`(Wednesday) and squashed the commits 👍 . Thanks @embano1 for the review 💯
